faster conversion between RGB and HSV colorspaces

grlee77 opened the pull request:
## Description

The current code for `rgb2hsv` and `hsv2rgb` is pretty inefficient. This PR uses Cython to accelerate these two functions, while also reducing the peak memory required. Related, to #1133, but this PR only improves two of the functions in the module.

For the included benchmarks, I get ~20 times faster computation and the memory used by for hsv2rgb is several times smaller. The cause of the large memory footprint in the current hsv2rgb implementation is the following lines where several temporary arrays matching the size of the input image are created:
https://github.com/scikit-image/scikit-image/blob/821c7f27f21c5fcf6a881a0640b94ebfd751665c/skimage/color/colorconv.py#L315-L322

## Note
The calls to `reshape(-1, 3)` are used to collapse multiple spatial dimensions into a single spatial axis to loop over in the Cython code. The output of the Cython function is then restored to the original shape.

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

grlee77 committed:
Cython implementation for rgb2hsv and hsv2rgb

grlee77 committed:
add benchmarks for rgb2hsv and hsv2rgb

benchmark both float32 and float64

pep8speaks commented:
Hello @grlee77! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`benchmarks/benchmark_color.py`](https://github.com/scikit-image/scikit-image/blob/ffcc1f44706a4d840610a975897b705af53db3f2/benchmarks/benchmark_color.py):

> [Line 45:80](https://github.com/scikit-image/scikit-image/blob/ffcc1f44706a4d840610a975897b705af53db3f2/benchmarks/benchmark_color.py#L45): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)

grlee77 commented:
I wrote simple CUDA kernel code for all color conversions in [cuCIM](https://github.com/rapidsai/cucim). This is done by JIT compiling elementwise kernels for each color conversion ([example for separate_stains](https://github.com/rapidsai/cucim/blob/600d6ab62d305daa53832b4ac99d1597e20a3644/python/cucim/src/cucim/skimage/color/colorconv.py#L1631-L1648)), giving a large acceleration over the CPU code.

The cases involving HSV were the ones where the relative performance difference was largest, so I have backported a CPU equivalent here for that one.

grlee77 commented:
Actually, these functions are also a reasonable use case for [Pythran](https://pythran.readthedocs.io/en/latest/). I tried this just out of interest and the following is what the Pythran code for `rgb2hsv_inner` that I came up with looks like:

```Python
#pythran export rgb2hsv_inner(float64[:, 3] order (C))
#pythran export rgb2hsv_inner(float32[:, 3] order (C))
def rgb2hsv_inner(rgb):
    hsv = np.empty_like(rgb)
    n = rgb.shape[0]
    for i in range(n):
        minv = rgb[i, :].min()
        maxv = rgb[i, :].max()
        delta = maxv - minv
        if delta == 0.0:
            hsv[i, :2] = 0.0
        else:
            hsv[i, 1] = delta / maxv
            if rgb[i, 0] == maxv:
                hsv[i, 0] = (rgb[i, 1] - rgb[i, 2]) / delta
            elif rgb[i, 1] == maxv:
                hsv[i, 0] = 2.0 + (rgb[i, 2] - rgb[i, 0]) / delta
            elif rgb[i, 2] == maxv:
                hsv[i, 0] = 4.0 + (rgb[i, 0] - rgb[i, 1]) / delta
            hsv[i, 0] /= 6.0
            hsv[i, 0] -= floor(hsv[i, 0])
        hsv[i, 2] = maxv
    return hsv
```

It is pretty similar to the Cython case, but a bit simpler. This Pythran version was actually slightly faster than the Cython code here on small images, but ~25% slower on large ones. I am not sure what is the underlying cause for the difference.

grlee77 committed:
fix copy-paste error in color/setup.py

rfezzani commented on skimage/color/_colorconv.pyx:
```diff
@@ -0,0 +1,81 @@
+        Py_ssize_t i, n, ch
+
+    n = rgb.shape[0]
+    for i in range(n):
```
What about `prange` here?

grlee77 replied:
I can try it. There may not be enough computation for it to provide a benefit, though.

grlee77 replied:
I tried `prange` and did see some improvements by a factor of 2-5 for 10 cores (20 threads) on large images. For some reason the `rgb2hsv` function became much slower than before if setting OMP_NUM_THREADS=1 before running the benchmarks though! That seems odd to me, but I wasn't able to quickly diagnose the source of the problem so I have left it without `prange` for now.

rfezzani reviewed (commented):
Thank you @grlee77 :wink:

rfezzani commented:
Pythran version loops twice over the channels looking for min and max value while in your implementation there is a unique traversal. This may explain the performance gain :wink:

grlee77 commented on skimage/color/colorconv.py:
```diff
@@ -225,48 +226,17 @@ def rgb2hsv(rgb):
+    if not rgb.flags.c_contiguous:
+        rgb = np.ascontiguousarray(rgb)
+    rgb = rgb.reshape(-1, 3)
+    hsv = rgb2hsv_inner(rgb).reshape(shape)
```
yeah, probably so. I could also switch back to `out`, but I think it is a bit clearer to just use `hsv`

rfezzani replied:
I have no strong preference :wink:

grlee77 commented:
> Pythran version loops twice over the channels looking for min and max value while in your implementation there is a unique traversal.

I thought that too, but tried changing it and didn't see much difference

grlee77 committed:
use nogil context

grlee77 committed:
update output variable name in docstring

rfezzani commented on skimage/color/_colorconv.pyx:
```diff
@@ -0,0 +1,83 @@
+                elif rgb[i, 2] == maxv:
+                    hsv[i, 0] = 4.0 + (rgb[i, 0] - rgb[i, 1]) / delta
+                hsv[i, 0] /= 6.0
+                hsv[i, 0] -= floor(<double>hsv[i, 0])
```
I am not sure that this will work but what about defining a `_floor` function specific to `np_floats`?
```
if np_floats is cnp.float32_t:
    _floor = floorf
else:
    _floor = floor
```
it may avoid this cast to double precision when the input is single precision...

grlee77 replied:
I think we would need to a C99 flag (or `language='c++'`) to ensure `floorf` is available. I think that used to be a problem for older MSVC versions, but is probably fine now. 

Cython 0.29.x doesn't define floorf in it's `libc.math`, but the current 3.0 alpha does. That doesn't really matter, though, since we can always just add it ourselves with: 
```Cython
cdef extern from "<math.h>" nogil:
    float floorf(float)
```

grlee77 committed:
Merge remote-tracking branch 'upstream/main' into color_rgb_hsv_cython

hmaarrfk commented on skimage/color/_colorconv.pyx:
```diff
@@ -0,0 +1,83 @@
+        Py_ssize_t i, n, hi, rem
+
+    n = rgb.shape[0]
+    with nogil:
```
Are you able to add a reference for the implementation of the algorithm here? That would allow future people to continue to optimize it.

grlee77 replied:
yes, I have added references in 0fab7ef

grlee77 committed:
More efficient modulus in rgb2hsv inner

Only a specific branch potentially needs to wrap negative values.

DOC: document the source of the algorithm and the publication where HSV was proposed

jonahpearl commented:
The skimage version is still pretty slow -- I tried just grabbing OP's suggestion but it doesn't seem any faster. Any suggested fixes? Or is the current skimage version actually a sped-up implementation, and it was even slower before? Thanks!

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
